python_modules/dagster/dagster/core/execution/context/compute.py

```diff
@@ -237,6 +245,31 @@ def get_tag(self, key: str) -> Optional[str]:
         """
         return self._step_execution_context.get_tag(key)
 
+    def has_events(self) -> bool:
+        return bool(self._events)
+
+    def retrieve_events(self) -> Iterator[DagsterEvent]:
+        while self._events:
+            yield self._events.pop(0)
+
+    def log_event(self, event) -> None:
```

Comment:
This still has the issue I raised on the last version, where asset materializations won't show up in the event log until the op completes, and might be out of order with log messages. Is that right?

Reply:
As I understand it, event logs are created at the time that a DagsterEvent is initialized, so by creating the DagsterEvent immediately, we preserve ordering in the event log. [1] attempts to test this by making sure that the timestamps showing up in the event log are accurate to when they actually occur, rather than when the DagsterEvent is yielded. Tl;dr no they should be in order

Reply:
It looks to me like `DagsterEvent`s don't have a timestamp on them. The timestamp is on `EventLogRecord`, which wraps the Dagster event and is created later. Am I reading right?

Reply:
It's on `EventLogEntry`, which is created immediately when the `DagsterEvent` is created

Reply:
Ahhhhhh. Wow, I find it very unexpected that we call `log_asset_event` inside `DagsterEvent.from_step`, etc., but in light of that, it looks like this code does that I was asking for.